### PR TITLE
Gradle 8 updates

### PR DESCRIPTION
**Pull request by shubham1g5**
cross-request: https://github.com/dimagi/commcare-core/pull/1277

### CODE REVIEW

**Comment by avazirna** [comment]
@shubham1g5 should we plan this for 2.55?

**Comment by shubham1g5** [comment]
> should we plan this for 2.55?

yes

**Comment by avazirna** [comment]
@damagatchi retest this please

**Comment by avazirna** [comment]
@damagatchi retest this please

**Comment by avazirna** [comment]
@damagatchi retest this please

**Comment by avazirna** [comment]
@damagatchi retest this please

File: app/test-proguard.cfg [hunk 1]
```diff
@@ -0,0 +1,15 @@
+-dontwarn com.sun.javadoc.**
```
Comment (shubham1g5) [inline app/test-proguard.cfg:1]: should the proguard rules for test be same as prod ?

File: app/proguard.cfg [hunk 2]
```diff
@@ -102,6 +104,7 @@
 -dontnote okhttp3.**
 
 #Retrofit Stuff
+-keep class retrofit2.** { *; }
```
Comment (shubham1g5) [inline app/proguard.cfg:107]: How does Gradle update affect proguard rules ?

Comment (avazirna) [reply app/proguard.cfg:107]: Gradle 8 is on [full mode](https://r8.googlesource.com/r8/+/refs/heads/master/compatibility-faq.md#r8-full-mode) by default, which means more aggressive optimizations. In this case, it was causing `IllegalArgumentExceptions` with the message `Call return type must be parameterized as Call<Foo> or Call<? extends Foo>` during get requests, which made me suspect that maybe some references had been obfuscated. After some digging I found [some posts](https://stackoverflow.com/questions/64439204/call-return-type-must-be-parameterized-as-callfoo-or-call-extends-foo-excep) that pointed that way too.

File: app/test-proguard.cfg (hunk 1 above)
Comment (avazirna) [reply app/test-proguard.cfg:1]: My understanding is that, in previous version, ProGuard rules were mostly applied to `production` apks, very little (if any) optimizations to test `apks`. Gradle 8 with R8 in fullMode by default changed that, so now we also need specific rules for the tests `apk`.

File: app/proguard.cfg (hunk 2 above)
Comment (shubham1g5) [reply app/proguard.cfg:107]: got it, although I don't think it's a good practice to retain all code from a package. Probably we should instead take the rules from [here](https://github.com/square/retrofit/blob/trunk/retrofit/src/main/resources/META-INF/proguard/retrofit2.pro)

File: app/test-proguard.cfg (hunk 1 above)
Comment (avazirna) [reply app/test-proguard.cfg:1]: from the `missing_rules` file after a failed build. Here's a snippet of the report:
```
Missing class com.sun.javadoc.Doc (referenced from: java.lang.String org.checkerframework.org.plumelib.options.OptionsDoclet.javadocToHtml(com.sun.javadoc.Doc) and 3 other contexts)
Missing class com.sun.javadoc.DocErrorReporter (referenced from: boolean org.checkerframework.org.plumelib.options.OptionsDoclet.validOptions(java.lang.String[][], com.sun.javadoc.DocErrorReporter))
Missing class com.sun.javadoc.FieldDoc (referenced from: void org.checkerframework.org.plumelib.options.OptionsDoclet.processEnumJavadoc(org.checkerframework.org.plumelib.options.Options$OptionInfo) and 1 other context)
Missing class com.sun.javadoc.RootDoc (referenced from: com.sun.javadoc.RootDoc org.checkerframework.org.plumelib.options.OptionsDoclet.root and 5 other contexts)
Missing class com.sun.javadoc.SeeTag (referenced from: java.lang.String org.checkerframework.org.plumelib.options.OptionsDoclet.javadocToHtml(com.sun.javadoc.Doc))
Missing class com.sun.javadoc.Tag (referenced from: java.lang.String org.checkerframework.org.plumelib.options.OptionsDoclet.javadocToHtml(com.sun.javadoc.Doc))
Missing class com.sun.source.tree.AnnotatedTypeTree (referenced from: java.lang.Boolean annotator.find.ASTPathCriterion$1.visitAnnotatedType(com.sun.source.tree.AnnotatedTypeTree, com.sun.source.tree.Tree) and 26 other contexts)
Missing class com.sun.source.tree.AnnotationTree (referenced from: boolean annotator.find.TreeFinder.alreadyPresent(com.sun.source.util.TreePath, annotator.find.Insertion) and 27 other contexts)
Missing class com.sun.source.tree.ArrayAccessTree (referenced from: com.sun.source.tree.Tree annotator.find.ASTPathCriterion.getNext(com.sun.source.tree.Tree, scenelib.annotations.io.ASTPath, int) and 28 other contexts)
Missing class com.sun.source.tree.ArrayTypeTree (referenced from: com.sun.source.tree.ArrayTypeTree org.checkerframework.dataflow.cfg.node.ArrayTypeNode.tree and 29 other contexts)
Missing class com.sun.source.tree.AssertTree (referenced from: com.sun.source.tree.Tree annotator.find.ASTPathCriterion.getNext(com.sun.source.tree.Tree, scenelib.annotations.io.ASTPath, int) and 13 other contexts)
Missing class com.sun.source.tree.AssignmentTree (referenced from: com.sun.source.tree.Tree annotator.find.ASTPathCriterion.getNext(com.sun.source.tree.Tree, scenelib.annotations.io.ASTPath, int) and 23 other contexts)
Missing class com.sun.source.tree.BinaryTree (referenced from: com.sun.source.tree.BinaryTree org.checkerframework.checker.interning.InterningVisitor$2.val$node and 98 other contexts)
```

**Comment by shubham1g5** [comment]
@avazirna checking if you are stuck on something with this ?

**Comment by avazirna** [comment]
> @avazirna checking if you are stuck on something with this ?

Nothing, all good. Let me just check why did the build fail.

**Review by avazirna — DISMISSED** [review]

**Comment by avazirna** [comment]
@damagatchi retest this please

File: app/proguard.cfg (hunk 2 above)
Comment (shubham1g5) [reply app/proguard.cfg:107]: looks like this is still pending.

Comment (avazirna) [reply app/proguard.cfg]: yes, just pushed that also

**Comment by avazirna** [comment]
@damagatchi retest this please

**Comment by avazirna** [comment]
@damagatchi retest this please

**Comment by avazirna** [comment]
@damagatchi retest this please

**Comment by shubham1g5** [comment]
Great, looks good to merge to me as well.

**Review by avazirna — APPROVED** [review]

**Pull request merged**